django_socio_grpc/decorators.py

```diff
@@ -79,3 +106,256 @@ def wrapper(function):
+                else func.__qualname__
+            )
+            # INFO - AM - 21/08/2024 - We use the qualname to be sure to have the right function. We could have use the sender be this would mean having the owner and the name saved in the instance of the grpc action
+            if f"{owner.__name__}.{name}" != func_qual_name:
+                return
+
+            # INFO - AM - 22/08/2024 - Set the default value for key_prefix, senders, invalidator_signals is not set
+            nonlocal key_prefix, senders, invalidator_signals
+            if not key_prefix:
+                key_prefix = f"{owner.__name__}-{name}"
+            if invalidator_signals is None:
+                invalidator_signals = (post_save, post_delete)
+            if senders is None and hasattr(owner, "queryset"):
+                senders = owner.queryset.model
+
+            # INFO - AM - 22/08/2024 - If no sender are specified and the service do not have a queryset we can't use the cache deleter. There is a warning displayed to the user
+            if senders is not None:
+                if not isinstance(senders, Iterable):
+                    senders = [senders]
+
+                # INFO - AM - 21/08/2024 - Once we have all the value we need we can connect the Model signals to the cache deleter
+                @receiver(invalidator_signals, weak=False)
+                def invalidate_cache(*args, **kwargs):
+                    if kwargs.get("sender") in senders:
+                        # INFO - AM - 01/08/2024 - To follow django cache_page signature cache is a string and not a cache instance. In this behavior we need the cache instance so we get it
+                        cache_instance = caches[cache] if cache else default_cache
+                        # INFO - AM - 01/08/2024 - For now this is only for Redis cache but this is important to support as it simplify the cache architecture to create if used
+                        if hasattr(cache, "delete_pattern"):
+                            cache_instance.delete_pattern(
+                                f"views.decorators.cache.cache_header.{key_prefix}.*"
+                            )
+                            cache_instance.delete_pattern(
+                                f"views.decorators.cache.cache_page.{key_prefix}.*"
+                            )
+                        else:
+                            cache_instance.clear()
+            else:
+                logger.warning(
+                    "You are using cache_endpoint_with_deleter without senders. If you don't need the auto deleter just use cache_endpoint decorator."
+                )
+
+        # INFO - AM - 22/08/2024 - http_to_grpc is a decorator but as it is returned from a decorator (to be able to access func) we need to call it directly to have the actual decorator
+        return http_to_grpc(
+            method_decorator(
+                cache_page(timeout, key_prefix=key_prefix, cache=cache), name="cache_page"
```

Comment:
I don't think the nonlocal will make this work, the functions are already resolved and returned. This could be done by modifying action.function = http_to_grpc(...(action.function)) when the signal is triggered